### PR TITLE
WIP: Planner

### PR DESCRIPTION
**Pull request by garunagiri**
A preliminary planner that buckets abilities into detection, hunt, and response tactics to make decisions on when to take which actions

### CODE REVIEW

File: app/response_planner.py [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+    def __init__(self, operation, planning_svc, stopping_conditions=()):
+        self.operation = operation
+        self.planning_svc = planning_svc
+        self.stopping_conditions = stopping_conditions
+        self.stopping_condition_met = False
+        self.processed_links = []
+        self.handled_detections = []    # stores completed links that are processed
+        self.severity = 0
+
+    async def execute(self, phase=1):
```
Comment (wbooth) [inline app/response_planner.py:12]: phase out phases

File: app/response_planner.py [hunk 2]
```diff
@@ -0,0 +1,41 @@
+class LogicalPlanner:
```
Comment (wbooth) [inline app/response_planner.py:1]: rename to ResponsePlanner or other

File: app/response_planner.py [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+            elif link.ability.tactic == 'response':
+                if any(uf not in self.handled_detections for uf in link.used):
+                    to_apply.append(link)
+                    detections_being_handled.append(uf for uf in link.used if uf not in self.detections_to_handle)
+        return to_apply, detections_being_handled
+
+    def process_completed_links(self):
+        for l in [link for link in self.operation.chain if link not in self.processed_links]:
+            for rel in l.relationships:
+                if rel.trait == 'operation.severity.modifier' and l.relationships.length() > 1:
```
Comment (wbooth) [inline app/response_planner.py:39]: drop > 1

**Comment by privateducky** [comment]
exciting stuff ;) looking forward to the final copy here

File: app/response_planner.py (hunk 2 above)
Comment (garunagiri) [reply app/response_planner.py:1]: I looked into this. Line 225 in c_operation.py requires that the planner have LogicalPlanner as the class name

**Comment by wbooth** [comment]
@garunagiri can we update or close this?

**Comment by garunagiri** [comment]
I'll be getting back on this as soon as https://github.com/mitre/caldera/pull/1585 is finished. If closing the PR for now is appropriate we can do that

**Comment by privateducky** [comment]
prop tip: @ArtificialErmine recently dropped in a new planner interface we can/should look at adopting here

**Comment by wbooth** [comment]
@garunagiri open back up when you want eyes on

**Pull request closed without merging**